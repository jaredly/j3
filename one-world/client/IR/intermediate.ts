// An intermediate representation

import { Nodes, Node, Style } from '../../shared/nodes';
import { ListDisplay, RenderInfo } from '../../shared/renderables';

type Format = Style;

const refStyle: Style = {
    fontWeight: 'bold',
    // type: 'text', style: { fontWeight: 'bold' }
};

export type IR =
    | { type: 'text'; text: string; style?: Style }
    | {
          type: 'vert';
          items: IR[];
          style?: Style;
          layout?: { tightFirst: number; pairs?: number[]; indent: number };
      }
    // | { type: 'squish'; item: IR; maxWidth: number }
    | { type: 'inline'; items: IR[]; style?: Style }
    | { type: 'horiz'; items: IR[]; style?: Style; wrap?: number } // number indicates indent amount
    | { type: 'indent'; item: IR; amount?: number; style?: Style }
    | { type: 'switch'; options: IR[] }
    | { type: 'loc'; loc: number }
    | { type: 'punct'; text: string; style?: Style };

export type Layout =
    | { type: 'horiz'; wrap?: number }
    | {
          type: 'vert';
          layout: {
              tightFirst: number;
              // This is funny, because it's a list of the LHS locs that shouldn't be squished.
              pairs?: number[];
              indent: number;
          };
      }
    | { type: 'pairs' }
    | { type: 'switch' };

export const nodeToIR = (
    node: Node,
    styles: Record<number, Format>,
    layouts: Record<number, Layout>,
    names: Record<string, Record<number, string>>,
): IR => {
    switch (node.type) {
        case 'array':
        case 'list':
        case 'record': {
            const lr = {
                array: '[]',
                list: '()',
                record: '{}',
            }[node.type];

            const l = layouts[node.loc] ?? { type: 'switch' };

            const items = node.items.map((loc): IR => ({ type: 'loc', loc }));

            switch (l.type) {
                case 'switch':
                    return {
                        type: 'horiz',
                        items: [
                            { type: 'punct', text: lr[0] },
                            {
                                type: 'switch',
                                options: [
                                    { type: 'horiz', items },
                                    {
                                        type: 'vert',
                                        items,
                                        layout: { tightFirst: 1, indent: 2 },
                                    },
                                ],
                            },
                            { type: 'punct', text: lr[1] },
                        ],
                    };
                case 'horiz':
                    return {
                        type: 'horiz',
                        wrap: l.wrap,
                        items: [
                            { type: 'punct', text: lr[0] },
                            ...items,
                            { type: 'punct', text: lr[1] },
                        ],
                    };
                case 'vert':
                    return {
                        type: 'horiz',
                        items: [
                            { type: 'punct', text: lr[0] },
                            { type: 'vert', items: items, layout: l.layout },
                            { type: 'punct', text: lr[1] },
                        ],
                    };
            }
            throw new Error('unknown layout');
        }

        // case 'rich-text':
        //     return { type: 'text', text: 'rich' };

        case 'id':
            return {
                type: 'text',
                text: node.ref
                    ? names[node.ref.toplevel][node.ref.loc]
                    : node.text,
                style: styles[node.loc] ?? node.ref ? refStyle : undefined,
            };

        case 'string': {
            const multi =
                node.first.includes('\n') ||
                node.templates.some((t) => t.suffix.includes('\n'));
            return {
                type: 'horiz',
                items: [
                    { type: 'loc', loc: node.tag },
                    { type: 'punct', text: '"' },
                    {
                        type: multi ? 'inline' : 'horiz',
                        style: multi
                            ? {
                                  background: 'red',
                              }
                            : undefined,
                        items: [
                            { type: 'text', text: node.first },
                            ...node.templates.flatMap((t): IR[] => [
                                { type: 'punct', text: '${' },
                                { type: 'loc', loc: t.expr },
                                { type: 'punct', text: '}' },
                                { type: 'text', text: t.suffix },
                            ]),
                        ],
                    },
                    { type: 'punct', text: '"' },
                ],
            };
        }

        case 'record-access':
            return {
                type: 'horiz',
                wrap: 2,
                items: [
                    { type: 'loc', loc: node.target },
                    ...node.items.map(
                        (loc): IR => ({
                            type: 'horiz',
                            items: [
                                { type: 'punct', text: '.' },
                                { type: 'loc', loc },
                            ],
                        }),
                    ),
                ],
            };
        case 'annot':
            return {
                type: 'horiz',
                items: [
                    { type: 'loc', loc: node.contents },
                    { type: 'punct', text: ':' },
                    { type: 'loc', loc: node.annot },
                ],
            };
        case 'comment':
        case 'spread':
            return {
                type: 'horiz',
                items: [
                    {
                        type: 'punct',
                        text: node.type === 'spread' ? '..' : ';',
                    },
                    { type: 'loc', loc: node.contents },
                ],
            };
    }
};
