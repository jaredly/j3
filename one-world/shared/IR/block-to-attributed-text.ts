import { splitGraphemes } from '../../../src/parse/splitGraphemes';
import { Path, serializePath, Style } from '../nodes';
import { applyFormats } from './format';
import { Block, BlockSource, blockSourceKey, inlineSize } from './ir-to-blocks';

const charLen = (text: string) => splitGraphemes(text).length;
const toChunk = (text: string, style?: Style): ABlock[0][0] => ({
    text,
    len: charLen(text),
    style,
});

const breakByLines = (line: ABlock[0]) => {
    const res: ABlock = [[]];
    line.forEach((item) => {
        if (item.text.includes('\n')) {
            const parts = item.text.split('\n');
            res[res.length - 1].push(toChunk(parts[0], item.style));
            for (let i = 1; i < parts.length; i++) {
                res.push([toChunk(parts[i], item.style)]);
            }
        } else {
            res[res.length - 1].push(item);
        }
    });
    return res;
};

const joinInline = (blocks: ABlock[]) => {
    const res: ABlock = [];
    blocks.forEach((block) => {
        if (res.length) {
            res[res.length - 1] = res[res.length - 1].slice();
            res[res.length - 1].push(...block[0]);
            res.push(...block.slice(1));
        } else {
            res.push(...block);
        }
    });
    return res;
};

export const justify = (text: ABlock) => {
    const lls = text.map((line) => line.reduce((t, c) => t + c.len, 0));
    const max = Math.max(...lls);
    text.forEach((line, i) => {
        if (lls[i] < max) {
            const wh = white(max - lls[i]);
            line.push({ text: wh, len: wh.length });
        }
    });
    return text;
};

const mergeStyle = (one?: Style, two?: Style) => {
    if (!one) return two;
    if (!two) return one;
    return { ...one, ...two };
};

export const blockFormat = (text: ABlock, style?: Style) => {
    if (!style) return text;
    return text.map((line) =>
        line.map((chunk) => ({
            ...chunk,
            style: mergeStyle(style, chunk.style),
        })),
    );
};

export const maxLength = <T extends { length: number }>(l: T[]) =>
    l.map((l) => l.length).reduce((a, b) => Math.max(a, b));

export const white = (num: number) => Array(num + 1).join(' ');

export const joinChunks = (chunks: ABlock[], pullLast = false): ABlock => {
    const maxLines = maxLength(chunks);
    if (maxLines === 1) return [chunks.flat(2)]; // all on one line
    chunks.forEach((group, i) => {
        for (let k = group.length; k < maxLines; k++) {
            if (pullLast && i === chunks.length - 1) {
                group.unshift([{ text: '', len: 0 }]);
            } else {
                group.push([{ text: '', len: 0 }]);
            }
        }
        if (i < chunks.length - 1) {
            const lls = group.map((line) =>
                line.reduce((l, i) => l + i.len, 0),
            );
            const maxLen = Math.max(...lls);
            group.forEach((len, i) => {
                if (lls[i] < maxLen) {
                    const nl = maxLen - lls[i];
                    group[i].push({
                        text: white(nl),
                        len: nl,
                    });
                }
            });
        }
    });
    const result: ABlock = [];
    for (let i = 0; i < maxLines; i++) {
        result.push(chunks.flatMap((l) => l[i]));
    }
    return result;
};

export type BlockEntry = {
    source: BlockSource;
    shape:
        | {
              type: 'inline';
              start: [number, number];
              end: [number, number];
              hbounds: [number, number]; // min / max X
          }
        | {
              type: 'block';
              start: [number, number];
              width: number;
              height: number;
          };
};

const applyX0 = (text: ABlock, x0: number) => {
    if (x0 === 0) return text;
    if (text.length === 1) return text;
    const indent = white(x0);
    text.forEach((line, i) => {
        if (i > 0) {
            line.unshift({ text: indent, len: x0 });
        }
    });
    return text;
};

export type RGB = { r: number; g: number; b: number };

export type StyleOverrides = Record<
    string,
    | {
          type: 'full';
          color: RGB;
      }
    | {
          type: 'sub';
          color: RGB;
          start: number;
          end: number;
      }
>;

export type DropTarget = {
    path: Path;
    pos: { x: number; y: number };
    side: 'before' | 'after';
    // after: { x: number; y: number };
};

// outer list is lines
// inner list is for different styles
type ABlock = {
    text: string;
    len: number;
    style?: Style;
}[][];

export const aBlockToString = (block: ABlock, color: boolean) => {
    return block
        .map((line) =>
            line
                .map((chunk) => applyFormats(chunk.text, chunk.style, color))
                .join(''),
        )
        .join('\n');
};

export type BlockCtx = {
    color: boolean;
    styles: StyleOverrides;
    sourceMaps?: BlockEntry[];
    dropTargets?: DropTarget[];
};

export type BlockPos = {
    x: number;
    y: number;
    x0: number;
};

export const blockToABlock = (
    pos: BlockPos,
    block: Block,
    ctx: BlockCtx,
): ABlock => {
    let nodeStyle = block.style;
    let override = false;
    if (block.node) {
        let key = serializePath(block.node);
        const brace = block.type === 'inline' && block.brace;
        if (brace) key += ':brace';
        const bstyle = ctx.styles[key];
        if (bstyle?.type === 'full') {
            override = true;
            nodeStyle = nodeStyle
                ? { ...nodeStyle, background: bstyle.color }
                : { background: bstyle.color };
        }
        if (ctx.dropTargets && !brace) {
            ctx.dropTargets.push({
                path: block.node,
                side: 'before',
                pos: { x: pos.x, y: pos.y },
            });
            if (block.width > 0) {
                ctx.dropTargets.push({
                    path: block.node,
                    side: 'after',
                    pos: {
                        x: pos.x + block.width,
                        y: pos.y + block.height - 1,
                    },
                });
            }
        }
    }

    switch (block.type) {
        case 'block': {
            if (block.horizontal !== false) {
                let x = pos.x;
                let h = 0;
                const chunks = block.contents.map((m, i) => {
                    h = Math.max(h, m.height);
                    let y = pos.y;
                    if (block.pullLast && i === block.contents.length - 1) {
                        y = pos.y + h - m.height;
                    }
                    const res = blockToABlock({ y, x, x0: x }, m, ctx);
                    x +=
                        m.width +
                        (i < block.contents.length - 1
                            ? (block.horizontal as number)
                            : 0);
                    return res;
                });
                if (block.horizontal > 0) {
                    const space = white(block.horizontal);
                    for (let i = chunks.length - 1; i >= 1; i--) {
                        chunks.splice(i, 0, [
                            [{ text: space, len: block.horizontal }],
                        ]);
                    }
                }
                let text = joinChunks(chunks, block.pullLast);
                return blockFormat(applyX0(text, pos.x - pos.x0), nodeStyle);
            }
            let y = pos.y;
            return blockFormat(
                applyX0(
                    block.contents
                        .map((m) => {
                            const res = blockToABlock(
                                { ...pos, y, x0: pos.x },
                                m,
                                ctx,
                            );
                            y += m.height;
                            return res;
                        })
                        .flat(),
                    pos.x - pos.x0,
                ),
                nodeStyle,
            );
        }
        case 'inline': {
            if (block.source && ctx.sourceMaps) {
                ctx.sourceMaps.push({
                    source: block.source,
                    shape:
                        block.height === 1
                            ? {
                                  type: 'block',
                                  start: [pos.x, pos.y],
                                  width: block.width,
                                  height: block.height,
                              }
                            : {
                                  type: 'inline',
                                  start: [pos.x, pos.y],
                                  end: [
                                      pos.x0 + block.last,
                                      pos.y + block.height - 1,
                                  ],
                                  hbounds: [
                                      pos.x0,
                                      Math.max(
                                          pos.x0 + block.width,
                                          pos.x + block.first,
                                      ),
                                  ],
                              },
                });
            }
            const override = block.source
                ? ctx.styles[blockSourceKey(block.source)]
                : undefined;
            if (typeof block.contents === 'string') {
                // if (block.contents.includes('\n')) {
                //     throw new Error('need tosplit newlines');
                // }
                const chars = splitGraphemes(block.contents);
                if (override?.type === 'sub') {
                    const pre = chars.slice(0, override.start);
                    const mid = chars.slice(override.start, override.end);
                    const post = chars.slice(override.end);
                    if (block.splits) {
                        for (let i = block.splits.length - 1; i >= 0; i--) {
                            const split = block.splits[i];
                            if (split > override.end) {
                                post.splice(split - override.end, 0, '\n');
                            } else if (split > override.start) {
                                mid.splice(split - override.start, 0, '\n');
                            } else {
                                pre.splice(split, 0, '\n');
                            }
                        }
                    }
                    // TODO: account for newlines in this text!!
                    return breakByLines([
                        {
                            text: pre.join(''),
                            len: pre.length,
                            style: nodeStyle,
                        },
                        {
                            text: mid.join(''),
                            len: mid.length,
                            style: {
                                ...nodeStyle,
                                background: override.color,
                            },
                        },
                        {
                            text: post.join(''),
                            len: post.length,
                            style: nodeStyle,
                        },
                    ]);
                }

                if (block.splits) {
                    for (let i = block.splits.length - 1; i >= 0; i--) {
                        const split = block.splits[i];
                        chars.splice(split, 0, '\n');
                    }
                }

                if (override?.type === 'full') {
                    return breakByLines([
                        {
                            text: chars.join(''),
                            len: chars.length,
                            style: {
                                ...nodeStyle,
                                background: override.color,
                            },
                        },
                    ]);
                }
                return breakByLines([
                    {
                        text: chars.join(''),
                        len: chars.length,
                        style: nodeStyle,
                    },
                ]);
            }
            const style: Style | undefined =
                override?.type === 'full'
                    ? { ...nodeStyle, background: override.color }
                    : nodeStyle;
            let x = pos.x;
            let y = pos.y;
            let x0 = pos.x0;
            return blockFormat(
                block.contents
                    .map((line, i) => {
                        const { width, height, positions, last } = inlineSize(
                            line,
                            x - x0,
                        );
                        const res = joinInline(
                            line.map((m, i) => {
                                const start = positions[i];
                                const res = blockToABlock(
                                    {
                                        x: x0 + start.x,
                                        y: y + start.y,
                                        x0: x0,
                                    },
                                    m,
                                    ctx,
                                );
                                if (m.type !== 'inline') {
                                    return justify(res);
                                }
                                return res;
                            }),
                        );

                        x = x0;
                        y += height;

                        return res;
                    })
                    .flat(1),
                style,
            );
        }
        case 'table':
            let y = pos.y;
            return blockFormat(
                block.rows
                    .map((row) => {
                        const chunks: ABlock[] = [];
                        let x = pos.x;
                        if (block.sides) {
                            chunks.push([[{ text: '⋮', len: 1 }]]);
                            x++;
                        }
                        let height = 0;
                        for (let i = 0; i < row.length; i++) {
                            if (i > 0) {
                                chunks.push([[{ text: '⋮', len: 1 }]]);
                                x++;
                            }
                            chunks.push(
                                blockToABlock(
                                    { ...pos, y, x0: x, x },
                                    row[i],
                                    ctx,
                                ),
                            );
                            if (row[i].width < block.colWidths[i]) {
                                const rest = block.colWidths[i] - row[i].width;
                                chunks.push([
                                    [{ text: white(rest), len: rest }],
                                ]);
                            }
                            x += block.colWidths[i];
                            height = Math.max(height, row[i].height);
                        }
                        if (block.sides) {
                            chunks.push([[{ text: '⋮', len: 1 }]]);
                            x++;
                        }
                        y += height;
                        return joinChunks(chunks);
                    })
                    .flat(1),
                nodeStyle,
            );
    }
};
