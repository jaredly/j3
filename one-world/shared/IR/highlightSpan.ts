import { splitGraphemes } from '../../../src/parse/splitGraphemes';
import { BlockEntry } from './block-to-text';
import { underlineText } from './format';

export const matchesSpan = (
    x: number,
    y: number,
    span: BlockEntry['shape'],
) => {
    if (span.type === 'block') {
        return (
            span.start[0] <= x &&
            span.start[0] + span.width > x &&
            span.start[1] <= y &&
            span.start[1] + span.height > y
        );
    }
    if (y === span.start[1]) {
        return x >= span.start[0] && x <= span.hbounds[1];
    }
    if (y === span.end[1]) {
        return x >= span.hbounds[0] && x <= span.end[0];
    }
    return (
        y > span.start[1] &&
        y < span.end[1] &&
        x >= span.hbounds[0] &&
        x <= span.hbounds[1]
    );
};

export const highlightSpan = (text: string, span: BlockEntry['shape']) => {
    const lines = text.split('\n');
    if (span.type === 'block') {
        return lines
            .map((line, i) => {
                if (
                    i >= span.start[1] &&
                    i <= span.start[1] + span.height - 1
                ) {
                    const chars = splitGraphemes(line);
                    if (span.width === 0 && span.height === 1) {
                        return (
                            chars.slice(0, span.start[0]).join('') +
                            '|' +
                            chars.slice(span.start[0]).join('')
                        );
                    }
                    return (
                        chars.slice(0, span.start[0]).join('') +
                        underlineText(
                            chars
                                .slice(
                                    span.start[0],
                                    span.start[0] + span.width,
                                )
                                .join(''),
                        ) +
                        chars.slice(span.start[0] + span.width).join('')
                    );
                }
                return line;
            })
            .join('\n');
    }
    return lines
        .map((line, i) => {
            if (i >= span.start[1] && i <= span.end[1]) {
                const chars = splitGraphemes(line);
                const start =
                    i === span.start[1] ? span.start[0] : span.hbounds[0];
                const end = i === span.end[1] ? span.end[0] : span.hbounds[1];
                // if (start > end) return line;
                return (
                    chars.slice(0, start).join('') +
                    underlineText(chars.slice(start, end).join('')) +
                    chars.slice(end).join('')
                );
            }
            return line;
        })
        .join('\n');
};
