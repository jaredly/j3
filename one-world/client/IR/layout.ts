import { Style } from '../../shared/nodes';
import { IR } from './intermediate';

// type LayoutCache = Record<number, any>
type IRForLoc = Record<number, IR>;

type LayoutResult = {
    maxWidth: number;
    inlineWidth: number;
    inlineHeight: number;
    height: number;
};

export const space = 8;

export type LayoutChoices = Record<
    number,
    | { type: 'hwrap'; groups: number[] }
    | { type: 'switch'; which: number }
    | { type: 'text-wrap'; splits: number[] }
>;

export const layoutIR = (
    x: number,
    dedent: number,
    ir: IR,
    choices: LayoutChoices,
    ctx: {
        maxWidth: number;
        // width for LHSs
        leftWidth: number;
        irs: IRForLoc;
        layouts: Record<
            number,
            { result: LayoutResult; choices: LayoutChoices }
        >;
        // layouts: LayoutCache,
        textLayout: (
            text: string,
            dedent: number,
            style?: Style,
        ) => LayoutResult;
    },
): LayoutResult => {
    switch (ir.type) {
        case 'switch': {
            let res: LayoutResult;
            let which = ir.options.length - 1;
            for (let i = 0; i < ir.options.length; i++) {
                res = layoutIR(x, dedent, ir.options[i], choices, ctx);
                if (res.maxWidth + x <= ctx.maxWidth) {
                    which = i;
                    break;
                    // ir.which = i;
                    // return res;
                }
            }
            choices[ir.id] = { type: 'switch', which };
            return res!;
        }

        case 'loc': {
            const choices: LayoutChoices = {};
            const result = layoutIR(x, dedent, ctx.irs[ir.loc], choices, ctx);
            ctx.layouts[ir.loc] = { result, choices };
            return result;
        }
        case 'punct':
        case 'text':
            return ctx.textLayout(ir.text, dedent, ir.style);

        case 'inline': {
            let maxWidth = 0;
            let inlineWidth = 0;
            let height = 0;
            let inlineHeight = 0;
            ir.items.forEach((item, i) => {
                let next = layoutIR(
                    x + inlineWidth,
                    inlineWidth,
                    item,
                    choices,
                    ctx,
                );
                if (next.inlineHeight < next.height) {
                    height += next.height - inlineHeight;
                    maxWidth = Math.max(maxWidth, inlineWidth);
                    // we're on a new line
                    inlineWidth = next.inlineWidth;
                    inlineHeight = next.inlineHeight;
                } else {
                    inlineHeight = Math.max(inlineHeight, next.inlineHeight);
                    inlineWidth += next.inlineWidth;
                }
            });
            return { maxWidth, inlineWidth, inlineHeight, height };
        }

        case 'squish':
            return layoutIR(x, 0, ir.item, choices, {
                ...ctx,
                maxWidth: ctx.leftWidth,
            });

        case 'vert': {
            let inlineWidth = 0;
            let inlineHeight = 0;
            let maxWidth = 0;
            let height = 0;

            ir.items.forEach((item, i) => {
                const next = layoutIR(x, 0, item, choices, ctx);
                inlineWidth = next.inlineWidth;
                inlineHeight = next.inlineHeight;
                maxWidth = Math.max(maxWidth, next.maxWidth);
                height += next.height;
            });

            return { inlineWidth, inlineHeight, maxWidth, height };
        }
        case 'horiz': {
            let lineWidth = 0;
            let lineHeight = 0;
            let maxWidth = 0;
            let height = 0;
            let inlineHeight = 0;
            const groups: number[] = [0];

            ir.items.forEach((item, i) => {
                const w = layoutIR(x + lineWidth, 0, item, choices, ctx);
                lineWidth += w.maxWidth;
                inlineHeight = Math.max(inlineHeight, w.inlineHeight);

                if (ir.wrap != null && lineWidth + x > ctx.maxWidth && i > 0) {
                    maxWidth = Math.max(maxWidth, lineWidth - w.maxWidth);
                    height += lineHeight;

                    groups.push(i);
                    const w2 = layoutIR(
                        x + ir.wrap.indent * space,
                        0,
                        item,
                        choices,
                        ctx,
                    );
                    lineWidth = ir.wrap.indent * space + w2.maxWidth;
                    lineHeight = w2.height;
                    inlineHeight = w2.inlineHeight;
                } else {
                    lineHeight = Math.max(lineHeight, w.height);
                }
            });

            if (ir.wrap != null) {
                choices[ir.wrap.id] = { type: 'hwrap', groups };
            }

            maxWidth = Math.max(maxWidth, lineWidth);
            height += lineHeight;

            return { maxWidth, inlineWidth: lineWidth, height, inlineHeight };
        }
        case 'indent': {
            const indent = space * (ir.amount ?? 1);
            const res = layoutIR(x + indent, 0, ir.item, choices, ctx);
            return {
                maxWidth: res.maxWidth + indent,
                inlineWidth: res.inlineWidth + indent,
                height: res.height,
                inlineHeight: res.inlineHeight,
            };
        }
    }
    // throw new Error('no');
};
