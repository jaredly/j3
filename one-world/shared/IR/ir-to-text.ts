// yay

import { splitGraphemes } from '../../../src/parse/splitGraphemes';
import { Style } from '../nodes';
import { IR, IRSelection } from './intermediate';
import { LayoutChoices, LayoutCtx } from './layout';

export const maxLength = <T extends { length: number }>(l: T[]) =>
    l.map((l) => l.length).reduce((a, b) => Math.max(a, b));

const white = (num: number) => Array(num + 1).join(' ');

export const joinChunks = (chunks: string[], pullLast = false) => {
    const lined = chunks.map((c) => c.split('\n'));
    const max = maxLength(lined); // lined.map(l => l.length).reduce((a, b) => Math.max(a, b))
    if (max === 1) return chunks.join('');
    lined.forEach((group, i) => {
        for (let k = group.length; k < max; k++) {
            if (pullLast && i === chunks.length - 1) {
                group.unshift('');
            } else {
                group.push('');
            }
        }
        if (i < lined.length - 1) {
            const ml = maxLength(group);
            group.forEach((line, i) => {
                if (line.length < ml) {
                    group[i] = line + white(ml - line.length);
                }
            });
        }
    });
    const result = [];
    for (let i = 0; i < max; i++) {
        result.push(lined.map((l) => l[i]).join(''));
    }
    return result.join('\n');
};

const addSpaces = (
    items: string[],
    mode: 'all' | 'start' | 'end',
    space = ' ',
) => {
    let min = mode === 'all' || mode === 'end' ? 1 : 2;
    let max = items.length - (mode === 'all' || mode === 'start' ? 1 : 2);
    for (let i = max; i >= min; i--) {
        items.splice(i, 0, space);
    }
};

export const irToText = (
    ir: IR,
    irs: Record<number, IR>,
    choices: LayoutChoices,
    layouts: LayoutCtx['layouts'],
    selection?: { path: number[]; sel: IRSelection; start?: IRSelection },
    space = ' ',
): string => {
    switch (ir.type) {
        case 'loc': {
            const { choices } = layouts[ir.loc];
            const sub =
                selection?.path[0] === ir.loc
                    ? { ...selection, path: selection.path.slice(1) }
                    : undefined;
            return irToText(irs[ir.loc], irs, choices, layouts, sub, space);
        }
        case 'cursor':
            if (selection?.path.length === 0) {
                if (
                    selection.sel.type === 'side' &&
                    selection.sel.side === ir.side
                ) {
                    return '|';
                }
                if (
                    selection.start?.type === 'side' &&
                    selection.start.side === ir.side
                ) {
                    return '.';
                }
            }
            return '';
        case 'control':
            switch (ir.control.type) {
                case 'check':
                    return ir.control.checked ? '[x] ' : '[ ] ';
                case 'radio':
                    return ir.control.checked ? '(x) ' : '( ) ';
                case 'bullet':
                    return ' - ';
                case 'number':
                    return `${ir.control.num
                        .toString()
                        .padStart(ir.control.width, ' ')}) `;
                default:
                    throw new Error('no');
            }
        case 'punct':
            return ir.text;
        case 'inline': {
            const wrap = choices[ir.wrap];
            if (!wrap || wrap.type !== 'hwrap')
                throw new Error('no wrap for inline ' + ir.wrap);

            const lines: string[][] = [];
            ir.items.forEach((item, i) => {
                if (wrap.groups.includes(i)) {
                    lines.push([]);
                }
                const last = lines[lines.length - 1];
                last.push(
                    irToText(item, irs, choices, layouts, selection, space),
                );
            });

            const pre = wrap.groups.length ? wrap.groups.join(',') + '\n' : '';

            return pre + lines.map((chunks, i) => chunks.join('')).join('!\n');
        }
        case 'horiz':
            const wrap = ir.wrap != null ? choices[ir.wrap.id] : null;
            if (ir.wrap && wrap?.type === 'hwrap') {
                // const pre = wrap.groups.join(',') + '\n';
                const lines: string[][] = [];
                ir.items.forEach((item, i) => {
                    if (wrap.groups.includes(i)) {
                        lines.push([]);
                    }
                    const last = lines[lines.length - 1];
                    last.push(
                        irToText(item, irs, choices, layouts, selection, space),
                    );
                });
                return lines
                    .map((chunks, i) => {
                        if (ir.spaced) {
                            addSpaces(chunks, 'all', space);
                        }
                        if (i > 0 && ir.wrap!.indent > 0) {
                            chunks.unshift(white(ir.wrap!.indent));
                        }
                        return joinChunks(
                            chunks,
                            ir.pullLast && i === lines.length - 1,
                        );
                    })
                    .join('\n');
            }
            const chunks = ir.items.map((item) =>
                irToText(item, irs, choices, layouts, selection, space),
            );
            if (ir.spaced) {
                addSpaces(chunks, 'all', space);
            }
            return joinChunks(chunks, ir.pullLast);
        case 'indent':
            return joinChunks([
                white(ir.amount ?? 2),
                irToText(ir.item, irs, choices, layouts, selection, space),
            ]);
        case 'squish':
            return irToText(ir.item, irs, choices, layouts, selection, space);
        case 'text': {
            const text = applyFormats(ir.text, ir.style);
            if (ir.wrap == null) return text;
            const splits = choices[ir.wrap];
            if (splits?.type === 'text-wrap' && splits.splits.length) {
                let pieces = [];
                for (let i = splits.splits.length - 1; i >= 0; i--) {
                    if (i === splits.splits.length - 1) {
                        pieces.unshift(text.slice(splits.splits[i]));
                    } else {
                        pieces.unshift(
                            text.slice(splits.splits[i], splits.splits[i + 1]),
                        );
                    }
                    if (i === 0) {
                        pieces.unshift(text.slice(0, splits.splits[0]));
                    }
                }
                return pieces.join('↩\n');
            }
            return text; // todo wrappp
        }
        case 'switch':
            const choice = choices[ir.id];
            if (choice?.type !== 'switch') {
                throw new Error(`switch not resolved ${ir.id}`);
            }
            return irToText(
                ir.options[choice.which],
                irs,
                choices,
                layouts,
                selection,
                space,
            );
        case 'vert':
            if (ir.pairs) {
                // ok I'm gonna say, if it's pairs, it's all pairs.
                const pairs: (
                    | { type: 'other'; item: string }
                    | { type: 'pair'; left: string; l: number; right: string }
                )[] = [];
                let lw = 0;
                ir.items.forEach((item) => {
                    if (item.type === 'horiz' && item.items.length === 2) {
                        const left = irToText(
                            item.items[0],
                            irs,
                            choices,
                            layouts,
                            selection,
                            space,
                        );
                        let l = maxLength(left.split('\n'));
                        lw = Math.max(lw, l);
                        pairs.push({
                            type: 'pair',
                            left,
                            l,
                            right: irToText(
                                item.items[1],
                                irs,
                                choices,
                                layouts,
                                selection,
                                space,
                            ),
                        });
                    } else {
                        pairs.push({
                            type: 'other',
                            item: irToText(
                                item,
                                irs,
                                choices,
                                layouts,
                                selection,
                                space,
                            ),
                        });
                    }
                });
                lw += 1;
                return pairs
                    .map((item) =>
                        item.type === 'other'
                            ? item.item
                            : joinChunks([
                                  item.l < lw
                                      ? item.left
                                            .split('\n')
                                            .map((line) =>
                                                line.length < lw
                                                    ? line +
                                                      white(lw - line.length)
                                                    : line,
                                            )
                                            .join('\n')
                                      : item.left,
                                  item.right,
                              ]),
                    )
                    .join('\n');
            }
            return ir.items
                .map((item) =>
                    irToText(item, irs, choices, layouts, selection, space),
                )
                .join('\n');
    }
};

const normal = 'abcdefghijklmnopqrstuvwxyz';
const bold = splitGraphemes('𝐚𝐛𝐜𝐝𝐞𝐟𝐠𝐡𝐢𝐣𝐤𝐥𝐦𝐧𝐨𝐩𝐪𝐫𝐬𝐭𝐮𝐯𝐰𝐱𝐲𝐳');
const italic = splitGraphemes('𝑎𝑏𝑐𝑑𝑒𝑓𝑔ℎ𝑖𝑗𝑘𝑙𝑚𝑛𝑜𝑝𝑞𝑟𝑠𝑡𝑢𝑣𝑤𝑥𝑦𝑧');
const bolditalic = splitGraphemes('𝒂𝒃𝒄𝒅𝒆𝒇𝒈𝒉𝒊𝒋𝒌𝒍𝒎𝒏𝒐𝒑𝒒𝒓𝒔𝒕𝒖𝒗𝒘𝒙𝒚𝒛');
const underline = String.fromCharCode(818);

const applyFormats = (text: string, style?: Style) => {
    if (!style) return text;
    if (style.fontWeight === 'bold') {
        if (style.fontStyle === 'italic') {
            text = convert(text, bolditalic);
        } else {
            text = convert(text, bold);
        }
    } else if (style.fontStyle === 'italic') {
        text = convert(text, italic);
    }
    if (style.textDecoration === 'underline') {
        const emes = splitGraphemes(text);
        text = emes.map((m) => m + underline).join('');
    }
    return text;
};

const convert = (text: string, alph: string[]) => {
    let res = '';
    for (let i = 0; i < text.length; i++) {
        const idx = normal.indexOf(text[i]);
        if (idx === -1) res += text[i];
        else res += alph[idx];
    }
    return res;
};
