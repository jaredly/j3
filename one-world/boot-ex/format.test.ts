// testingi t up
import { expect, test } from 'bun:test';
import { joinChunks } from '../shared/IR/ir-to-text';

test('joinChunks', () => {
    expect(joinChunks(['a', 'b'])).toBe('ab');
    expect(joinChunks(['a\nc', 'b'])).toBe('ab\nc');
    expect(joinChunks(['a\ncd', 'b'])).toBe('a b\ncd');
    expect(joinChunks(['a\ncd', 'b', 'abc\na'])).toBe('a babc\ncd a');
});
