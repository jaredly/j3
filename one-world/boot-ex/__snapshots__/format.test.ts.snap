// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`basic wrap 1`] = `
"
(hello,folks,lets
   see,how,this,goes
   and,such,like
   that)
"
`;

exports[`let + pairs 1`] = `
"
(let,[x              y
      a              b
      (one,two,three 6
         four,five)   ]
   one
   two
   three               )
"
`;

exports[`nesty nest 1`] = `
"
(one,two
   (three,four,five
      size)
   seven,eight,nine)
"
`;

exports[`smol wrap 1`] = `
"
(one,two,seven
   eight,nine
   ten,eleven
   twelve
   thirteen
   fourteen)
"
`;

exports[`stringsss 1`] = `
"
"Here is a string"
"
`;

exports[`stringsss 2`] = `
"
"Here ${is}a string"
"
`;

exports[`stringsss nl 1`] = `
"
"Here is a
 string ${a}with things
 and such. ${(
    one,two,three)}you know."
"
`;

exports[`string long one 1`] = `
"
"what in the world ↩
 here we are it is ↩
 good somewhat ↩
 things-that tend-to-↩
 be splittable ↩
 insufficiently       "
"
`;

exports[`long string with inclusions 1`] = `
"
"Here is a ${here}we are it is good ↩
 somewhat things-↩
 that ${tend-to-be}splittable ↩
 insufficiently                      "
"
`;

exports[`gradual wrap 1`] = `
"
------------|
(abc,def,ghi
   a)
-----------|
(abc,def,ghi
   a)
----------|
(abc,def
   ghi,a)
---------|
(abc,def
   ghi,a)
--------|
(abc,def
   ghi,a)
-------|
(abc,def
   ghi,a)
"
`;
